File: Corte1/C++/Parcial/Vista.cpp
#ifndef VISTA_CPP
#define VISTA_CPP
#include <iostream>
#include <iomanip>
#include <string>
#include "GestorCurso.cpp"

using namespace std;

class Vista {
private:
    GestorCurso& gestor;

    // VALIDACIÓN DE ROBUSTEZ: Evita que el programa se rompa si ingresan letras
    int leerEntero(string mensaje) {
        int valor;
        cout << mensaje;
        while (!(cin >> valor)) {
            cout << ">> [!] ERROR: Debe ingresar solo numeros: ";
            cin.clear();
            cin.ignore(1000, '\n');
        }
        return valor;
    }

public:
    Vista(GestorCurso& g) : gestor(g) {}

    // REQUERIMIENTO 23 y 45: Menú Principal
    void ejecutar() {
        int op = 0;
        cout << "=== SISTEMA DE GESTION ACADEMICA ===" << endl;
        while (op != 4) {
            cout << "\n1. Registrar Alumno\n2. Calificar Alumno\n3. Ver Reporte\n4. Salir\n\nOpcion: ";
            op = leerEntero("");

            if (op == 1) capturar();
            else if (op == 2) calificar();
            else if (op == 3) reporte();
        }
    }

private:
    void capturar() {
        string c, n;
        cout << "Codigo: "; cin >> c;
        cout << "Nombre: "; cin.ignore(); getline(cin, n);
        int e = leerEntero("Edad: ");
        if(gestor.registrar(Estudiante(c, n, e))) cout << ">> Alumno guardado.\n";
        else cout << ">> [!] Memoria llena.\n";
    }

    void calificar() {
        string c;
        cout << "Ingrese el codigo del estudiante: "; cin >> c;
        int idx = gestor.buscar(c);

        // MENSAJE DE ERROR: Estudiante no encontrado
        if(idx == -1) {
            cout << ">> [!] ERROR: El estudiante con codigo (" << c << ") no esta registrado." << endl;
        } else {
            cout << "Calificando a: " << gestor.getEst(idx).getNombre() << endl;
            for(int j=0; j < MAX_NOTAS; j++) {
                float v;
                cout << "Nota " << j+1 << " (0-5): ";
                while(!(cin >> v) || v < 0 || v > 5) {
                    cout << "Error. Use numeros entre 0 y 5: ";
                    cin.clear(); cin.ignore(1000, '\n');
                }
                gestor.setNota(idx, j, v);
            }
        }
    }

    void reporte() {
        if(gestor.getCont() == 0) {
            cout << ">> No hay datos registrados." << endl;
            return;
        }
        cout << "\n" << left << setw(15) << "ESTUDIANTE" << " | " << "PROMEDIO" << endl;
        cout << "-----------------------------------" << endl;
        for(int i=0; i < gestor.getCont(); i++) {
            cout << left << setw(15) << gestor.getEst(i).getNombre() 
                 << " | " << fixed << setprecision(2) << gestor.getPromedio(i) << endl;
        }
    }
};
#endif

File: Corte2/Python/Evaluación/Vista.py
from Estudiante import Estudiante
from Configuracion import MAX_NOTAS

class Vista:
    def __init__(self, g):
        self.__gestor = g

    # VALIDACIÓN DE ROBUSTEZ: Evita que el programa se rompa si ingresan letras
    def __leer_entero(self, mensaje):
        while True:
            try:
                valor = int(input(mensaje))
                return valor
            except ValueError:
                print(">> ERROR: Debe ingresar solo numeros: ", end="")

    # REQUERIMIENTO 23 y 45: Menú Principal
    def ejecutar(self):
        op = 0
        print("=== SISTEMA DE GESTION ACADEMICA ===")
        while op != 4:
            print("\n1. Registrar Alumno\n2. Calificar Alumno\n3. Ver Reporte\n4. Salir")
            op = self.__leer_entero("\nOpcion: ")

            if op == 1: self.__capturar()
            elif op == 2: self.__calificar()
            elif op == 3: self.__reporte()

    def __capturar(self):
        c = input("Codigo: ")
        n = input("Nombre: ")
        e = self.__leer_entero("Edad: ")
        if self.__gestor.registrar(Estudiante(c, n, e)):
            print(">> Alumno guardado.")
        else:
            print(">> Memoria llena.")

    def __calificar(self):
        c = input("Ingrese el codigo del estudiante: ")
        idx = self.__gestor.buscar(c)

        # MENSAJE DE ERROR: Estudiante no encontrado
        if idx == -1:
            print(f">> ERROR: El estudiante con codigo ({c}) no esta registrado.")
        else:
            print(f"Calificando a: {self.__gestor.get_est(idx).get_nombre()}")
            for j in range(MAX_NOTAS):
                while True:
                    try:
                        v = float(input(f"Nota {j+1} (0-5): "))
                        if 0 <= v <= 5:
                            self.__gestor.set_nota(idx, j, v)
                            break
                        else:
                            print("Error. Use numeros entre 0 y 5: ", end="")
                    except ValueError:
                        print("Error. Ingrese un valor numerico: ", end="")

    def __reporte(self):
        if self.__gestor.get_cont() == 0:
            print(">> No hay datos registrados.")
            return
        
        print(f"\n{'ESTUDIANTE':<15} | {'PROMEDIO'}")
        print("-" * 35)
        for i in range(self.__gestor.get_cont()):
            nombre = self.__gestor.get_est(i).get_nombre()
            promedio = self.__gestor.get_promedio(i)
            print(f"{nombre:<15} | {promedio:.2f}")